File: 20-10-14-ap-war/python/__main__.py
import unittest

# See the README in the parent directory for the description and instructions
# for this activity.


def under_vs_upper(under: str, upper: str) -> str:
    # Write your implementation here...
    pass


# Below are tests!
class TestAPWar(unittest.TestCase):
    under_victory = "Underclassmen are victorious"
    upper_victory = "Upperclassmen are victorious"
    tie = "No victor on this battlefield"
    correct = ["Expected under victory",
               "Expected upper victory", "Expected tie"]

    def test_under_wins(self):
        self.run_test("0 0 0 0 10", "0 0 0 0 0",
                      self.under_victory, self.correct[0])
        self.run_test("1 0 1 0 0", "0 1 0 0 0",
                      self.under_victory, self.correct[0])

    def test_upper_wins(self):
        self.run_test("0 1 0 0 1", "0 0 0 1 1",
                      self.upper_victory, self.correct[1])
        self.run_test("0 0 0 0 0", "0 0 0 0 10",
                      self.upper_victory, self.correct[1])

    def test_tie(self):
        self.run_test("0 0 1 0 1", "1 0 1 1 0", self.tie, self.correct[2])
        self.run_test("2 0 4 0 5", "6 2 2 0 3", self.tie, self.correct[2])
        self.run_test("0 0 0 0 10", "0 0 0 0 10", self.tie, self.correct[2])

    def run_test(self, under, upper, answer, message):
        self.assertEqual(under_vs_upper(under, upper),
                         answer, message)


if __name__ == '__main__':
    unittest.main()

File: 20-10-14-ap-war/java/src/Main.java
public class Main {
  // Implement this method!
  public static String underVsUpper(String under, String upper) {
    // Replace this with your implementation...
    throw new UnsupportedOperationException("Not yet implemented!");
  }



  // Below are tests...
  public static void main(String[] args) {
    Tester tester = new Tester();
    tester.test();
  }

  static class Tester {

    private boolean noErrors = true;

    // See test cases:
    private void testUnderWins() {
      runTest("under1", "1 0 1 0 0", "0 1 0 0 0", UNDER_VICTORY);
      runTest("under2", "0 0 0 0 10", "0 0 0 0 0", UNDER_VICTORY);
    }

    private void testUpperWins() {
      runTest("upper1", "0 1 0 0 1", "0 0 0 1 1", UPPER_VICTORY);
      runTest("upper2" , "0 0 0 0 0", "0 0 0 0 10", UPPER_VICTORY);
    }

    private void testTie() {
      runTest("tie1", "0 0 1 0 1", "1 0 1 1 0", TIE);
      runTest("tie2", "2 0 4 0 5", "6 2 2 0 3", TIE);
      runTest("tie3", "0 0 0 0 10", "0 0 0 0 10", TIE);
    }


    private static final String UNDER_VICTORY = "Underclassmen are victorious";
    private static final String UPPER_VICTORY = "Upperclassmen are victorious";
    private static final String TIE = "No victor on this battlefield";

    public void test() {
      System.out.println("Running tests...");

      testUnderWins();
      testUpperWins();
      testTie();

      System.out.println("Finished running tests!");
      if (this.noErrors) {
        System.out.println("All tests passed!");
      } else {
        System.out.println("Some tests failed.");
      }
    }

    private void runTest(String id, String under, String upper, String answer) {
      try {
        String result = Main.underVsUpper(under, upper);
        if (!result.equals(answer)) {
          printTestError(id, "Assertion failed: " + result + " != " + answer);
        }
      } catch (Exception e) {
        printTestError(id, "Exception caught: " + e);
      }
    }

    private void printTestError(String id, String message) {
      if (this.noErrors) {
        System.err.println();
        this.noErrors = false;
      }
      System.err.println("Test '" + id + "'\n... " + message + "\n");
    }
  }
}
